### ISSUE TITLE
`put` should not insert versions directly - they should only come from `check`

### ISSUE THREAD

**Issue by vito** [concourse/concourse#3463]
## What challenge are you facing?

see https://github.com/concourse/git-resource/issues/18, https://github.com/concourse/git-resource/issues/196, https://github.com/concourse/git-resource/issues/221

This is already planned along with concourse/rfcs#1, but I'm gonna log an explicit issue for it as it has user-facing impact.

As of 5.0 we have decoupled build inputs and outputs from resource versions which means we don't have to directly add any entries to the version history - they'll be correlated by the `version_md5` once they're discovered by `check`. Having `put` insert directly into the resource versions table breaks expectations for users who currently use filtering rules on the resource in order to limit `check`, because if they perform a `put` to that same resource with a version that *normally* would have been ignored, the version still gets discovered because the `put` directly inserts it.

## What would make this better?

Stop doing that! Just record it as an output and move on.

This will be backwards-incompatible for any resources that rely on this behavior - for example by having a no-op `check` and a `put` that directly adds versions that propagate to later builds. I think that's OK and a necessary change as it's not really what resources are designed for. For additional context, this was discussed as part of concourse/rfcs#1.

Also related: concourse/rfcs#11, which also would solve this issue by allowing the versions to be recorded but have Concourse know not to trigger any builds for'em.

**Comment by JohannesRudolph** [comment]
Sure - let me break my answer down into two parts because I have another question. What is the implication of this for the "implict get after put" workflow? 

Not stating that I was a particular fan of the implicit get by default but there's probably a lot of pipelines out there that rely on this. These will need a way to migrate to something else. I see two useful features that will be lost/need replacement: 

-`put` emits a new version without requiring a separate `check`, which means that concourse can start triggering other dependent jobs that `get` this resource with `trigger: true` immediately. This is great for reducing pipeline latency.
  - sidenote: the implicit get isn't "useless" from a global perspective because at least one worker will already now have a resource input version volume cached
- `put` with implicit `get` provides a way to generate a new resource version in a job and then immediately continue using that resource version in the same job.  If `put` no longer emits a new version `X`, the only way to work with `X` would be to have a separate `get` step. But how do we guarantee this step fetches `X` and not some other version?
- Related to the previous point, how does another job state a `passed` constraint on a resource emitted by a previous job? How do we achieve safe fan-in semantics (i.e. get me `X` emitted by the previous job)?

Tbh. as a very interested outsider with quite an investment in concourse it's not easy to follow the evolution of the RFCs around resources, i.e. the current one appears to be https://github.com/concourse/rfcs/pull/37 -> supersedes https://github.com/concourse/rfcs/pull/24 -> supersedes https://github.com/concourse/rfcs/pull/1. I'm sorry if my questions are already clarified somewhere in those RFCs or in the discussion that lead to them. I'm sure a lot of thought has gone into those RFCs but maybe you can view my questions as sort of a litmus test of "end user concerns"  that a potential migration guide for the major version bump has to answer ;-)

**Comment by JohannesRudolph** [comment]
> Would you mind explaining more about it the conditional triggers flow? I'd like to see how it squares with RFCs we have in flight. (Thanks!)

Sure, prepare for a longer read :-) Let me preface that our primary build pipeline is more complex than I'd like it to be. I hope Concourse will provide better support for our use case in the future. I'd thus be very happy to answer any questions you have about it and would also be available for a call to clarify things.

## Context

We use Concourse for > 4 years. Our software is developed as a mono-repo, but contains many different subsystems (think: services). We run one primary build pipeline on concourse and many (>10) deploy pipelines. We've gone through quite a few evolutions of our primary build pipeline. The one we currently have is very solid, but also quite complex and certainly an "advanced" use of Concourse.

## Problem statement

We use a PR based workflow using GitHub. We want each PR to be built & fully validated by CI. Most branches/features only touch 1-2 subsystems at a time. It would be too expensive to simply rebuild each subsystem on every feature branch push or develop/master branch push.

However, we need to publish a "release" at the end of the pipeline. A release contains build artifacts from all subsystems. Before the release, we also run `qa` tests that need build artifacts from some subsystems.

You may quickly realise that this is sort of a "caching"/control-flow problem at the end of the day.

### Example

Say our mono repo has git commit `G1` and subsystems `sA` and `sB`. We have a release `R1` built from `G1` that includes `R1@G1={sA@G1, sB@G1}`. Now we have a change `G2` that only affects subsystem `sA` so we need to build a release `R2@G2={sA@G2, sB@G1}`.

## Job Graphs

### Subsystem Job Graph Layout

Each subsystem build has it's own "job graph" that typically consists of `build & unit test` and `acceptance` jobs. These job graphs can be different for each subsystem.

For example, **acceptance** tests are "slow" and prone to flaky/intermittent failures. It's useful to be able to **retry** them from the Concourse UI by retriggering the job build. This can "unstuck" a pipeline waiting on a failed acceptance test. That's why job-graphs can have individual acceptance jobs.

Seperating these into jobs also gives a better visual representation of commits flowing through the pipeline.

Sidenotes: Nowadays We use GitHub commit check status more frequently to determine PR build progress rather than Concourse UI. Concourse of course drives GitHub commit checks using [Cogito](https://github.com/Pix4D/cogito). There's too much builds going on in a pipeline to be useful to a single developer anyway.

### Control Flow as a Resource

Since we use a mono-repo, there is only a single logical git repo triggering our pipeline. We want to model **build control flow** that skips building unnecessary job graphs, yet is able to fan-in all required build artifacts to run qa and build release at the end of the pipeline.

There are challenges modeling control-flow in Concourse:

- Concourse has no notion of "skipping" a **job**. See [https://discuss.concourse-ci.org/t/trigger-jobs-conditionally/306](https://discuss.concourse-ci.org/t/trigger-jobs-conditionally/306) for a discussion
- Concourse also has very limited support of "skipping" **tasks**, i.e. you can use a `do` sequence nested in `try` and then fail a task in the `do` sequence to skip all subsequent tasks
- Concourse is conceptually built around "steady state" not a "workflow"

[gate-resource](https://github.com/meshcloud/gate-resource) allows us to have separate "branches" in the job-graph for each subsystem. Using the gate-resource allows us to model "dependencies" in our build pipeline explicitly and fan in results from disparate job-graphs into new pipelines (e.g. autoclose gates trigger jobs after each of the dependent gates have been closed). We also use this heavily to model things like "release channels" (i.e. "customer A receives beta releases", "customer B only receives stable releases").

### The "init" job

We have an `init` job in our pipeline that's triggered from a `git-branch-heads` resource for all PR/develop/master commits. This job determines which subsystems needs to be built. This is essentially based on a `subsystem-revs` script that determines subsystem revisions based on `git tree`...

Each subsystem has its own gate backed by gate-resource, with gate items like `sA@G1`, `sA@G2` etc. indicating that a job-graph for a subsystem has completed successfully.
The init job uses the `try do` trick to conditionally trigger [keyval-resource](https://github.com/SWCE/keyval-resource) to "trigger" subsystem job-graphs. The last job of the job-graph is then responsible for putting the appropriate gate item, e.g. `sA@G1`.

The init job writes [autoclose gate](https://github.com/meshcloud/gate-resource#auto-gates) files for qa and release jobs. You can think of these as "specifications" for the contents of a release. For example if init determines that release `R2@G2={sA@G2, sB@G1}` it writes an autoclose item `R2@G2` to the `release` gate that depends on `sA@G2, sB@G1`.

## Discarded Alternatives

At multiple points we thought we must be doing this all wrong and explored alternatives.

### Idempotent pipeline that always runs through all the jobs

This idea essentially means discarding job-graphs and having all commits run through a single pipeline.

This would require logic in each job / task to determine if it can be skipped. This is however not possible because the `try do` trick to skip tasks cannot be combined with reliable `on_error` handlers that correctly report commit status to GitHub because it's impossible to determine if the job was legitimately skipped or if it failed due to a build error.

### Mono Repo Buildsystems

We could also have an idempotent pipeline if we had a build-system doing all the caching. Build systems built for mono-repos like bazel can do that. However there's some serious downsides:

- we need a build system that's capable to run in a cluster to meet our performance/resource requirements
  - that means operating another cluster and we're already busy enough maintaining concourse on K8s/GKE
- we'd need to retool all subsystem builds to work with the new "mono" buildsystem
  - this means an investment on the same order of what we already invested in concourse
  - we loose the simplicity of using the "platform native" build systems like gradle/npm/go build for each subsystem/service
- bazel wasn't even 1.0 when we built our initial pipeline around gate-resource

And btw. we need concourse and gate-resource anyway to drive our delivery pipelines.

### "Per-branch" pipelines

When the initial "spaces" concept was discussed in https://github.com/concourse/rfcs/pull/1 we also looked into how that may solve our challenges.

However, 'm no longer sure that having a separate "PR pipeline" that is different from your "master" pipeline is even a good idea, cf. [Not Rocket Science Rule](https://graydon2.dreamwidth.org/1597.html). We use automated tooling to rebase/merge PRs ([kodiak](https://kodiakhq.com/)) and one nice property of our current build pipeline is that merging a PR **does not trigger any new builds**. Before the PR is even mergeable, all artifacts were built and tests ran.

The only thing that's done after PR merge is that a release gate is written, autoclosed, release gets assembled and the release is put in the right channel (based on whether the PR was merged to develop or master).

## Problems with Gate Resource / Room for Improvement

### Learning Curve 

Like concourse, gate resource is incredibly versatile because it's based on clean concepts, but applying it to a real-world use case requires a steep learning curve. It's simple once you "got it" but until then it's hard.

This is certainly an area where we can improve with better documentation and bettter output. This should help onboard new engineers understand it and its behavior better.

### Resource Fan In

Using gates means we lose concourse built-in fan-in of resouce versions into our jobs. That's a real pity because that means we need to manually fetch resource versions from backing stores in some places (e.g. GCS/S3 for build artifacts). It would be really cool if there was a dynamic resource interface where a task could say "hey resource X can you fetch me version V1 please" and concourse does that. 

## Impact of "`put` should not insert versions"

Some of the design decisions in gate-resource **are directly impacted by this change to put version behavior**.

The gate-resource is already designed so that it also correctly discovers versions using `check` which is really important for robustness in light of intermittent failures. So it's not one of the `put only` resources but instead it has proper check behavior.

However, we often setup dedicated "autoclose jobs" so that concourse immediately sees new versions emitted from autoclosing a gate and does not need to `check` the gate resource again and can start dependent jobs with minimum latency. To that end, instead of emitting 1 versions (currently possible) we'd want to even emit **multiple** versions in one go, see https://github.com/concourse/concourse/issues/2660

**Comment by vito** [comment]
Thanks a ton for the writeup! Not done processing it yet but I have at least *some* useful things to say so far so I figure I'll submit them early.

It sounds like your usage pattern has a lot of overlap with https://github.com/concourse/concourse/discussions/4949 - in particular https://github.com/concourse/concourse/discussions/4949#discussioncomment-5589 seems similar to your 'skipping' idea.

> Sure - let me break my answer down into two parts because I have another question. What is the implication of this for the "implict get after put" workflow?

This shouldn't have any effect on the `get` after a `put`, since in that case the version directly goes from one step to the next without relying on the version history at all. This would however have an impact on passing that version to downstream jobs; that would stop working.

> Tbh. as a very interested outsider with quite an investment in concourse it's not easy to follow the evolution of the RFCs around resources, i.e. the current one appears to be [concourse/rfcs#37](https://github.com/concourse/rfcs/pull/37) -> supersedes [concourse/rfcs#24](https://github.com/concourse/rfcs/pull/24) -> supersedes [concourse/rfcs#1](https://github.com/concourse/rfcs/pull/1). I'm sorry if my questions are already clarified somewhere in those RFCs or in the discussion that lead to them. I'm sure a lot of thought has gone into those RFCs but maybe you can view my questions as sort of a litmus test of "end user concerns" that a potential migration guide for the major version bump has to answer ;-)

Yeah, it can be kind of hard to follow through all the iterations. concourse/rfcs#37 is actually lower-level - concourse/rfcs#38 is what you want for the resource-specific use of it and can be considered a more direct replacement of concourse/rfcs#1.

**Comment by clarafu** [comment]
Just a quick note that when we get to implementing this, we currently only check for resources that are used as an input to a job within `LIDAR` so if a `put` step no longer inserts version and relies on a check of the resource to insert the version, we will either want a check to be triggered after the `put` step finishes or `LIDAR` to check resources that are used as both outputs and inputs.

https://github.com/concourse/concourse/blob/becd8ab072ccaa6cdb513e73dd8e7dbe69c384f2/atc/db/check_factory.go#L337-L341

**Comment by Rob4579** [comment]
Hi everyone!
Is there an actual ETA on that?  Also, what is the suggested workaround to temporary fix that issue?

Thank you! You guys are doing awesome work ;)

**Comment by vito** [comment]
@Rob4579 We're pretty hesitant to do this since it'll be a breaking change. I think the safest thing to do would be to implement this behavior for [Prototype-based resources](https://github.com/concourse/rfcs/blob/master/038-resource-prototypes/proposal.md) and preserve the existing behavior for today's resource types. That way we don't break any existing workflows, and we can phase out this behavior as the community gradually adopts Prototypes.

That would put us on a timeline for [Q4 2020](https://github.com/concourse/concourse/projects/90#card-39577022).

**Comment by Rob4579** [comment]
@vito That's perfect! 
Thanks you

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
